addon/globalPlugins/webAccess/gui/__init__.py

```diff
@@ -203,17 +258,229 @@ def initData(self, context):
+
+	def initData(self, context):
+		super().initData(context)
+		self.categoryClasses = self.initCatClasses()
+		self.catListCtrl.addToListCtrl(self.categoryClasses)
+		self.catListCtrl.SelectItem(self.catListCtrl.GetFirstChild(self.root)[0])
+		self.catListCtrl.ExpandAll()
+
+	def getFirstChild(self):
+		return self.categoryClasses[0]
+
+	def initCatClasses(self):
+		categoryClasses = []
+		for categoryClass, childrenGetterName in self.categoryInitList:
+			childrenGetter = getattr(self, childrenGetterName)
+			categoryClasses.append(
+				TreeNodeInfo(categoryClass, childrenGetter=childrenGetter, title=categoryClass.title))
+		return categoryClasses
+
+	def focus_first_field(self, evt):
+		# if evt.GetKeyCode() == wx.WXK_CONTROL_A:
+		page = self.tree.GetPage(self.tree.GetSelection())
+		children = page.GetChildren()
+		for child in children:
+			if child.__class__.__name__ in ['ComboBox', 'TextCtrl', 'Button']:
+				child.SetFocus()
+
+	def _changeCategoryPanel(self, newCatInfos):
+		configuredSettingsDialogType()
+		panel = self.catIdToInstanceMap.get(newCatInfos.title, None)
+		if panel:
+			panel.initData(self.context, **newCatInfos.categoryParams)
+			return panel
+		panel = newCatInfos.categoryClass(parent=self.container)
+		panel.Hide()
+		self.containerSizer.Add(
+			panel, flag=wx.ALL | wx.EXPAND,
+			border=guiHelper.SPACE_BETWEEN_ASSOCIATED_CONTROL_HORIZONTAL
+		)
+		self.catIdToInstanceMap[newCatInfos.title] = panel
+		panelWidth = panel.Size[0]
+		availableWidth = self.containerSizer.GetSize()[0]
+		if panelWidth > availableWidth : # and _isDebug():
+			log.debugWarning(
+				(
+						"Panel width ({1}) too large for: {0} Try to reduce the width of this panel, or increase width of " +
+						"MultiCategorySettingsDialog.MIN_SIZE"
+				).format(newCatInfos.categoryClass, panel.Size[0])
+			)
+		panel.initData(self.context, **newCatInfos.categoryParams)
+		panel.SetLabel(panel.title.replace('&', ""))
+		panel.SetAccessible(SettingsPanelAccessible(panel))
+		return panel
+
+	def _doCategoryChange(self, newCatInfos):
+		oldCat = self.currentCategory
+
+		# Freeze and Thaw are called to stop visual artifact's while the GUI
+		# is being rebuilt. Without this, the controls can sometimes be seen being
+		# added.
+		self.container.Freeze()
+		if oldCat:
+			oldCat.onPanelDeactivated()
+		try:
+			newPanel = self._changeCategoryPanel(newCatInfos)
+		except ValueError as e:
+			log.error("Unable to change to category: {}".format(newCatInfos.title), exc_info=e)
+			return
+		self.currentCategory = newPanel
+
+		newPanel.onPanelActivated()
+		# call Layout and SetupScrolling on the container to make sure that the controls apear in their expected locations.
+		self.container.Layout()
+		self.container.SetupScrolling()
+		self.container.Thaw()
+
+	def onCategoryChange(self, evt):
+		currentCat = self.currentCategory
+		newCatInfos = self.catListCtrl.getTreeNodeInfo(evt.GetItem())
+		if not currentCat or newCatInfos != currentCat:
```

Comment:
@GatienBouyssou
This test cannot fail: `currentCat` is here a `TreeContextualPanel`, not a `TreeNodeInfo`

Reply:
My bad thanks for noticing it. I removed the whole condition I don't think we have a use for it anymore. I don't see a case where any of these two conditions can be useful. currentCat is checked in _doCategoryChange and I don't think that onCategoryChange can be called twice for the same category. Please correct me if I'm wrong

Reply:
Sounds good to me.
Thanks